hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/replication/TestQuasiClosedStuckReplicaCount.java

```diff
@@ -0,0 +1,350 @@
+    Set<ContainerReplica> replicas = ReplicationTestUtil.createReplicasWithOriginAndOpState(
+        ContainerID.valueOf(1), QUASI_CLOSED,
+        Pair.of(origin1, IN_SERVICE), Pair.of(origin1, IN_SERVICE), Pair.of(origin1, IN_SERVICE),
+        Pair.of(origin1, DECOMMISSIONED));
+
+    QuasiClosedStuckReplicaCount replicaCount = new QuasiClosedStuckReplicaCount(replicas, 1);
+    assertFalse(replicaCount.isUnderReplicated());
+    assertFalse(replicaCount.isOverReplicated());
+  }
+
+  @Test
+  public void testUnderReplicationWithMaintenanceWithOneOrigin() {
+    Set<ContainerReplica> replicas = ReplicationTestUtil.createReplicasWithOriginAndOpState(
+        ContainerID.valueOf(1), QUASI_CLOSED,
+        Pair.of(origin1, IN_SERVICE), Pair.of(origin1, IN_SERVICE), Pair.of(origin1, ENTERING_MAINTENANCE));
+
+    QuasiClosedStuckReplicaCount replicaCount = new QuasiClosedStuckReplicaCount(replicas, 1);
+    assertFalse(replicaCount.isUnderReplicated());
+    assertFalse(replicaCount.isOverReplicated());
+
+    replicas = ReplicationTestUtil.createReplicasWithOriginAndOpState(
+        ContainerID.valueOf(1), QUASI_CLOSED,
+        Pair.of(origin1, IN_SERVICE), Pair.of(origin1, ENTERING_MAINTENANCE), Pair.of(origin1, ENTERING_MAINTENANCE));
+
+    replicaCount = new QuasiClosedStuckReplicaCount(replicas, 2);
+    assertTrue(replicaCount.isUnderReplicated());
+    assertFalse(replicaCount.isOverReplicated());
+    validateMisReplicatedOrigins(replicaCount.getUnderReplicatedReplicas(), 1, 3, 1, origin1);
+  }
+
+  @Test
+  public void testUnderReplicationWithMaintenanceWithTwoOrigins() {
+    Set<ContainerReplica> replicas = ReplicationTestUtil.createReplicasWithOriginAndOpState(
+        ContainerID.valueOf(1), QUASI_CLOSED,
+        Pair.of(origin1, IN_SERVICE), Pair.of(origin1, ENTERING_MAINTENANCE),
+        Pair.of(origin2, IN_SERVICE), Pair.of(origin2, IN_SERVICE));
+
+    QuasiClosedStuckReplicaCount replicaCount = new QuasiClosedStuckReplicaCount(replicas, 1);
+    assertFalse(replicaCount.isUnderReplicated());
+    assertFalse(replicaCount.isOverReplicated());
+
+    replicas = ReplicationTestUtil.createReplicasWithOriginAndOpState(
+        ContainerID.valueOf(1), QUASI_CLOSED,
+        Pair.of(origin1, ENTERING_MAINTENANCE), Pair.of(origin1, ENTERING_MAINTENANCE),
+        Pair.of(origin2, IN_SERVICE), Pair.of(origin2, IN_SERVICE));
+
+    replicaCount = new QuasiClosedStuckReplicaCount(replicas, 1);
+    assertTrue(replicaCount.isUnderReplicated());
+    assertFalse(replicaCount.isOverReplicated());
+    validateMisReplicatedOrigins(replicaCount.getUnderReplicatedReplicas(), 1, 2, 1, origin1);
+  }
+
+  @Test
+  public void testNoOverReplicationWithExcessMaintenanceReplicasTwoOrigins() {
+    Set<ContainerReplica> replicas = ReplicationTestUtil.createReplicasWithOriginAndOpState(
+        ContainerID.valueOf(1), QUASI_CLOSED,
+        Pair.of(origin1, IN_SERVICE), Pair.of(origin1, IN_SERVICE), Pair.of(origin1, IN_MAINTENANCE),
```

Comment:
Shouldn't this count as over replication? Min healthy for maintenance is one. So for origin1, one in-service replica and one maintenance replica should be sufficient.